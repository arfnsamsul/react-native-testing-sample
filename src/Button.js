import React, { Component } from 'react';
import { Text, TouchableOpacity, Linking } from 'react-native';

// 1. Changed to a class based component
class Button extends Component {
    constructor(props) {
        super(props);
    }

    // 2. Custom function called onPress TouchableOpacity
    onPressHandler = () => {
        const { onPress, url } = this.props
        if (url) {
            Linking.openURL(url)
        }
        onPress()
    }

    render(){
        const { buttonStyle, textStyle } = styles;
        const { label, primary } = this.props;

        // 3. Change color of button depending on 'primary' prop
        const newButtonStyle = primary ? buttonStyle : [buttonStyle, { backgroundColor: '#f34541', borderBottomColor: '#a43532' }];

        return (
            <TouchableOpacity onPress={this.onPressHandler} style={newButtonStyle}>
                <Text style={textStyle}>
                    {label}
                </Text>
            </TouchableOpacity>
        );
    }
};

Button.defaultProps = {
    primary: true,
}

const styles = {
    textStyle: {
        alignSelf: 'center',
        color: '#fff',
        fontSize: 16,
        fontWeight: '600',
    },
    buttonStyle: {
        height: 45,
        alignSelf: 'stretch',
        justifyContent: 'center',
        backgroundColor: '#38ba7d',
        borderBottomWidth: 6,
        borderBottomColor: '#1e6343',
        borderWidth: 1,
        marginLeft: 15,
        marginRight: 15
    }
};

export default Button;
