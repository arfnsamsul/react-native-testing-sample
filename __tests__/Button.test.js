import React from 'react';
import {shallow} from 'enzyme';
import Button from '../src/Button';

describe('Button', () => {
    describe('Rendering', () => {
        it('should match to snapshot - Primary', () => {
            const component = shallow(<Button label="test label" />)
            expect(component).toMatchSnapshot("Primary button snapshot")
        });
        it('should match to snapshot - Secondary', () => {
            const component = shallow(<Button label="test label" primary={false} />)
            expect(component).toMatchSnapshot("Secondary button snapshot")
        });
    });

    describe('Interaction', () => {
        describe('onPressHandler', () => {
            it('should call onPress', () => {
                // Arrange
                const mockOnPress = jest.fn();      // 1. mock function
                const component = shallow(<Button 
                    label= "test label" 
                    onPress={mockOnPress}           // 2. passing in mock function as props
                />);
                const instance = component.instance();  // 3. getting an instance of component

                // Act
                instance.onPressHandler();          // 4. manually triggering onPressHandler()

                // Assert
                expect(mockOnPress).toHaveBeenCalled();
                expect(mockOnPress).toHaveBeenCalledTimes(1);   // 5. checking return values
            });
        });
    });

});