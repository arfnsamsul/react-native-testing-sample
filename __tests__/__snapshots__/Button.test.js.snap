// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Rendering should match to snapshot - Primary: Primary button snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    label="test label"
    primary={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        test label
      </Text>,
      "onPress": [Function],
      "style": Object {
        "alignSelf": "stretch",
        "backgroundColor": "#38ba7d",
        "borderBottomColor": "#1e6343",
        "borderBottomWidth": 6,
        "borderWidth": 1,
        "height": 45,
        "justifyContent": "center",
        "marginLeft": 15,
        "marginRight": 15,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "test label",
        "style": Object {
          "alignSelf": "center",
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "600",
        },
      },
      "ref": null,
      "rendered": "test label",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#fff",
              "fontSize": 16,
              "fontWeight": "600",
            }
          }
        >
          test label
        </Text>,
        "onPress": [Function],
        "style": Object {
          "alignSelf": "stretch",
          "backgroundColor": "#38ba7d",
          "borderBottomColor": "#1e6343",
          "borderBottomWidth": 6,
          "borderWidth": 1,
          "height": 45,
          "justifyContent": "center",
          "marginLeft": 15,
          "marginRight": 15,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "test label",
          "style": Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          },
        },
        "ref": null,
        "rendered": "test label",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Button Rendering should match to snapshot - Secondary: Secondary button snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    label="test label"
    primary={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        test label
      </Text>,
      "onPress": [Function],
      "style": Array [
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#38ba7d",
          "borderBottomColor": "#1e6343",
          "borderBottomWidth": 6,
          "borderWidth": 1,
          "height": 45,
          "justifyContent": "center",
          "marginLeft": 15,
          "marginRight": 15,
        },
        Object {
          "backgroundColor": "#f34541",
          "borderBottomColor": "#a43532",
        },
      ],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "test label",
        "style": Object {
          "alignSelf": "center",
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "600",
        },
      },
      "ref": null,
      "rendered": "test label",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#fff",
              "fontSize": 16,
              "fontWeight": "600",
            }
          }
        >
          test label
        </Text>,
        "onPress": [Function],
        "style": Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#38ba7d",
            "borderBottomColor": "#1e6343",
            "borderBottomWidth": 6,
            "borderWidth": 1,
            "height": 45,
            "justifyContent": "center",
            "marginLeft": 15,
            "marginRight": 15,
          },
          Object {
            "backgroundColor": "#f34541",
            "borderBottomColor": "#a43532",
          },
        ],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "test label",
          "style": Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          },
        },
        "ref": null,
        "rendered": "test label",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
